dashboard/app/models/concerns/levels/levels_within_levels.rb

```diff
@@ -0,0 +1,59 @@
+# The LevelsWithinLevels module encapsulates all functionality related to the
+# various ways we have of combining levels, including:
+#
+# - BubbleChoice
+# - Contained Levels
+# - LevelGroup
+# - Project Templates
+module Levels
+  module LevelsWithinLevels
+    extend ActiveSupport::Concern
+    included do
+      # Because this parent-child relationship comes in different flavors, we
+      # also want to provide some scopes to make interacting with subsets of
+      # relationships easier.
+      module ByKindExtension
+        ParentLevelsChildLevel::VALID_KINDS.each do |kind|
+          define_method(kind) do
+            where(parent_levels_child_levels: {kind: kind})
+          end
+        end
+      end
+
+      # We store parent-child relationships in a self-referential join table.
+      # In order to define a has_many/through relationship in both directions,
+      # we must define two separate associations to the same join table.
+      has_many :levels_parent_levels,
+        class_name: 'ParentLevelsChildLevel',
+        foreign_key: :child_level_id
+      has_many :parent_levels,
+        -> {extending ByKindExtension},
+        through: :levels_parent_levels,
+        inverse_of: :child_levels
+
+      has_many :levels_child_levels,
+        class_name: 'ParentLevelsChildLevel',
+        foreign_key: :parent_level_id
+      has_many :child_levels,
+        -> {extending ByKindExtension},
+        through: :levels_child_levels,
+        inverse_of: :parent_levels
+    end
+
+    # Returns all child levels of this level, which could include contained
+    # levels, project template levels, BubbleChoice sublevels, or LevelGroup
+    # sublevels.  This method may be overridden by subclasses.
+    def all_child_levels
+      (contained_levels + [project_template_level] - [self]).compact
+    end
+
+    # we must search recursively for child levels, because some bubble choice
+    # sublevels have project template levels.
+    def all_descendant_levels
+      my_child_levels = all_child_levels
+      child_descendant_levels = my_child_levels.
+        map(&:all_descendant_levels).flatten
+      my_child_levels + child_descendant_levels
+    end
+  end
+end
```

Comment:
What are your thoughts on the effectiveness of extracting concerns versus extracting query or policy objects as a way of reducing model complexity?

I don't have a strong objection to extracting this concern, but it does seem to come at at cost of increasing the number of places the reader will have to look to find all the functionality of the including class. `all_child_levels` could be particularly tricky to track down because one finding its definition in the BubbleChoice model might reasonably expect to find the original definition in the Level model. this has also tripped me up when I've gone looking for "user permissions" in the user model, which have been extracted to a concern.

Given how big the level.rb file is, I'd rather have this extraction than not have it, especially since there is no corresponding query or policy object extraction that could be done in its place. but in general my take is that I'd prefer to rely on the observer pattern or other patterns which let us extract query or policy objects as a way to chip away at large model files.

Reply:
I definitely prefer POROs to concerns when what's being extracted is constrained to methods that can be explicitly invoked, but in this case the core functionality of what's being extracted is very closely tied to the association directives themselves. I considered keeping the associations on the model and extracting the methods to an object, but concluded that having those two halves of this functionality be separated from each other was more confusing than using a concern to put them all in the same place.

Reply:
Makes sense. Thanks for thinking through this!